test/test_cuda.py

```diff
@@ -3809,6 +3809,72 @@ def test_graph_make_graphed_callables(self):
+    def test_amp_graph_make_graphed_callables(self):
+        seed = 123
+        torch.manual_seed(seed)
+        torch.cuda.manual_seed(seed)
+
+        N = 20
+        C_in, L_in, C_out, L_out = 16, 50, 16, 14
+        kernel_size = 3
+
+        models = []
+        for _ in range(2):
+            model_section = torch.nn.Sequential(torch.nn.Linear(L_in, C_in, bias=False),
+                                                torch.nn.Conv1d(C_in, C_out, kernel_size, bias=False),
+                                                torch.nn.Sigmoid()).cuda()
+            models.append(model_section)
+
+        x = torch.randn(N, C_in, L_in, device='cuda', dtype=torch.float32)
+        y_pred = torch.randn(N, C_out, L_out, device='cuda', dtype=torch.float32, requires_grad=True)
+        y = torch.randn(N, C_out, L_out, device='cuda', dtype=torch.float32)
+
+        loss_fn_control = torch.nn.functional.mse_loss
+
+        for cache_enabled in (False, True,):
+            model_graphed = models[0]
+            model_control = models[1]
+
+            model_graphed.load_state_dict(model_control.state_dict())
+
+            opt_graphed = torch.optim.SGD(model_graphed.parameters(), lr=0.1)
+            opt_control = torch.optim.SGD(model_control.parameters(), lr=0.1)
+
+            with torch.cuda.amp.autocast(cache_enabled=cache_enabled):
+                model_graphed, loss_fn_graphed = \
+                    torch.cuda.make_graphed_callables((model_graphed, loss_fn_control),
+                                                      ((x,), (y_pred, y)))
+
+            real_inputs = [torch.rand_like(x) for _ in range(10)]
+            real_targets = [torch.rand_like(y) for _ in range(10)]
+
+
+            for m, opt, loss_fn in zip((model_graphed, model_control),
+                                       (opt_graphed, opt_control),
+                                       (loss_fn_graphed, loss_fn_control)):
+                # Resets RNC states before iterations for graphed and ungraphed models,
+                # so dropout math should be bitwise identical for both.
+                torch.manual_seed(seed)
+                torch.cuda.manual_seed(seed)
+                for data, target in zip(real_inputs, real_targets):
+                    opt.zero_grad(set_to_none=True)
+                    with torch.cuda.amp.autocast(cache_enabled=cache_enabled):
+                        y_pred = m(data)
+                        loss = loss_fn(y_pred, target)
+                        loss.backward()
+                    opt.step()
+
+            for pg, pc in zip(model_graphed.parameters(), model_control.parameters()):
+                self.assertEqual(pg, pc)
+
+            # We graphed the models in training mode. Eval should still run ungraphed.
+            model_graphed.eval()
+            model_control.eval()
+            self.assertEqual(model_graphed(real_inputs[0]), model_control(real_inputs[0]), exact_dtype=False)
```

Comment:
I am using `exact_dtype=False` here because `model_control` returns results in `torch.float32` while`model_graphed` in `torch.float16` because its graph was created in amp.autocast mode.